Review for DM-2659 (Psf class for great3sims modified to use the new library format)

jdswinbank opened the pull request:
This also requires the library building change in galaxy_shear_experiments

pgee2000 committed:
Psf class for great3sims modified to use the new library format

This also requires the library building change in galaxy_shear_experiments

jdswinbank commented on great3sims/psf.py:
```diff
@@ -149,7 +149,7 @@ class ConstPSFBuilder(PSFBuilder):
     # For lam/diam, Euclid has ~0.1375".
     #               WFIRST has a range for the different bands, but with 2.4m mirror and 1.1 micron,
     #               it is 0.0945".
-    # In terms of sampling, we use the correspondence between OpticalPSF <-> Airy 
```
Throughout, there are a lot of trivial whitespace changes. I agree that they are an improvement, but it's not appropriate to include them on the same commit as "real" work.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1382,10 +1382,22 @@ class DrawPSFBuilder(PSFBuilder):
     def __init__(self, obs_type, multiepoch, shear_type, draw_psf_src):
         self.obs_type = obs_type
         self.draw_psf_src = os.path.abspath(draw_psf_src)
+        filename = os.path.basename(self.draw_psf_src)
```
`filename` is unused.

jdswinbank replied:
I think you're substantially changing the semantics of the `DrawPSFBuilder` class here. Sadly the original doesn't seem to be overly blessed with documentation (at least, not in this file), however, for starters, it looks as though it used to assume `draw_psf_src` was a FITS file, and now you have it as a plain text catalogue.

This is not wrong as such, but it makes me wonder whether this is the best approach. Are you certain you will never need to use the original `DrawPSFBuilder` interface? Might your changes confuse others who have to work with this code in future? Did you consider subclassing rather than simply replacing the original?

pgee2000 replied:
This code was written by me exclusively for this project.  And as our approach to providing access to the Psfs has changed, so has the code.  I don't anticipate having a version of the code again which does not have an index, and no one else is relying on it.

We are not planning to push this code back to great3, but I agree that it could use a few more line comments, so I will add them.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1382,10 +1382,22 @@ class DrawPSFBuilder(PSFBuilder):
+        #   is the numeric index of the library, the "PsfIndex", which is the hdu index in the mini-
+        #   library, and the "PsfNumber", which is the index of the star in the PhoSim pars file
+        self.psfInfos = []
+        while True:
```
Again, it would be much nicer to use a more idiomatic loop, perhaps with a context manager for the open file:

``` python
with open(self.draw_psf_src, "r") as fin:
    for line in fin:
        entry = line.split()
        # etc
```

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1382,10 +1382,22 @@ class DrawPSFBuilder(PSFBuilder):
+        #   The input file draw_psf_src is a file which contains a list of all of the available Psfs
+        #   It contains the name of the mini-library which contains the psf, the "GroupIndex" which
+        #   is the numeric index of the library, the "PsfIndex", which is the hdu index in the mini-
+        #   library, and the "PsfNumber", which is the index of the star in the PhoSim pars file
```
Again, it's trivial, but the extra indentation in many of your comment blocks is weird.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1382,10 +1382,22 @@ class DrawPSFBuilder(PSFBuilder):
+            entry = fin.readline().split()
+            if len(entry) == 0:
+                break
+            self.psfInfos.append([entry[0], int(entry[1]), int(entry[1]), int(entry[2]), False])
```
What is `False`?

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+            record["psf_group"] = psfInfo[1]
+            record["psf_index"] = psfInfo[2]
+            record["psf_number"] = psfInfo[3]
+            psfInfo[4] = True
```
So I'm guessing `psfInfo[4]` is a boolean indicating whether this particular PSF was randomly selected. It'd be nice to document that somewhere.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+        dirname = os.path.dirname(self.draw_psf_src)
+        currentPsfFile = None
+        for i in range(len(self.psfInfos)):
+            psfInfo = self.psfInfos[i]
```
Prefer:

``` python
for i, psfInfo in enumerate(self.psfInfos):
    # do stuff
```

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+                        hdus.close()
+                        del hdus
+                    currentPsfFile = psfInfo[0]
+                    filename = dirname + "/" + psfInfo[0]
```
`os.path.join()`.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+                        del hdus
+                    currentPsfFile = psfInfo[0]
+                    filename = dirname + "/" + psfInfo[0]
+                    print "opening ", filename
```
I'm guessing this `print`, and the one two lines below, are just left over from debugging? Be nice to remove them.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+        for i in range(len(self.psfInfos)):
+            psfInfo = self.psfInfos[i]
+            if psfInfo[4]:
+                if psfInfo[0] != currentPsfFile:
```
Do you need all this logic involving `currentPsfFile`? It seems like you could probably avoid that variable altogether and do a simpler check based on comparing `psfInfo[0]` to `hdus.filename()`.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
-                psf.header["GS_SCALE"] = 0.02
-                psf.writeto(filename)
-        if hdus: hdus.close()
+                psf.header["GS_SCALE"] = 0.2
```
We wrote a `GS_SCALE` parameter in `makePsfLibrary.py`. Shouldn't we just be copying that over instead of redefining it? What happens if you later decide to change it, but forget that it's defined in (at least) two places.

Again, I think a comment describing what `GS_SCALE` is and why you have chosen 0.2 would be appropriate (I note it was originally 0.02 in the above, which makes me think this is doubly necessary).

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+                psf.header["CUNIT2"] = "deg"
+                psf.header["CRPIX1"] = 0
+                psf.header["CRPIX2"] = 0
+                psf.header["CD1_1"] = 5.55e-05
```
Why `5.55e-05`?

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+                psf.header["CD1_2"] = 0.0
+                psf.header["CD2_1"] = 0.0
+                psf.header["CD2_2"] = 5.55e-05
+                format = dirname + "/psf_%d.fits"
```
`os.path.join()`.

jdswinbank commented on great3sims/psf.py:
```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
+
         d = {
             'type':'InterpolatedImage',
+            'scale': .2,
```
Is this the same as `GS_SCALE`? If so, it should be defined in one place and re-used, rather than having `0.2`s popping up in various parts of the code.

pgee2000 committed:
Review comments

Also, added a change to the library naming to be compatible with makePsfLibrary
in galaxy_shear_experiments
